test/unit/solver/test_cuda_rosenbrock.cpp

```diff
@@ -0,0 +1,205 @@
+  auto jacobian = cpu_solver.GetState().jacobian_;
+  EXPECT_GE(jacobian.AsVector().size(), 13 * number_of_grid_cells);
+  for (auto& elem : jacobian.AsVector())
+    elem = 100.0;  
+  for (std::size_t i_cell = 0; i_cell < number_of_grid_cells; ++i_cell)
+  {
+    jacobian[i_cell][0][0] = 12.2;
+    jacobian[i_cell][0][1] = 24.3 * (i_cell + 2);
+    jacobian[i_cell][0][2] = 42.3;
+    jacobian[i_cell][1][0] = 0.43;
+    jacobian[i_cell][1][1] = 23.4;
+    jacobian[i_cell][1][2] = 83.4 / (i_cell + 3);
+    jacobian[i_cell][2][0] = 4.74;
+    jacobian[i_cell][2][2] = 6.91;
+    jacobian[i_cell][3][1] = 59.1;
+    jacobian[i_cell][3][3] = 83.4;
+    jacobian[i_cell][4][0] = 78.5;
+    jacobian[i_cell][4][2] = 53.6;
+    jacobian[i_cell][4][4] = 1.0;
+  }
+  cpu_solver.AlphaMinusJacobian(jacobian, 42.042);
+  return jacobian.AsVector(); 
+}
+
+
+template<template<class> class MatrixPolicy, template<class> class SparseMatrixPolicy, class LinearSolverPolicy>
+void testAlphaMinusJacobian(std::size_t number_of_grid_cells)
+{
+  auto solver = getSolver<MatrixPolicy, SparseMatrixPolicy, LinearSolverPolicy>(number_of_grid_cells);
+  auto jacobian = solver.GetState().jacobian_;
+  
+  EXPECT_EQ(jacobian.size(), number_of_grid_cells);
+  EXPECT_EQ(jacobian[0].size(), 5);
+  EXPECT_EQ(jacobian[0][0].size(), 5);
+  EXPECT_GE(jacobian.AsVector().size(), 13 * number_of_grid_cells);
+  for (auto& elem : jacobian.AsVector())
+    elem = 100.0;
+  for (std::size_t i_cell = 0; i_cell < number_of_grid_cells; ++i_cell)
+  {
+    jacobian[i_cell][0][0] = 12.2;
+    jacobian[i_cell][0][1] = 24.3 * (i_cell + 2);
+    jacobian[i_cell][0][2] = 42.3;
+    jacobian[i_cell][1][0] = 0.43;
+    jacobian[i_cell][1][1] = 23.4;
+    jacobian[i_cell][1][2] = 83.4 / (i_cell + 3);
+    jacobian[i_cell][2][0] = 4.74;
+    jacobian[i_cell][2][2] = 6.91;
+    jacobian[i_cell][3][1] = 59.1;
+    jacobian[i_cell][3][3] = 83.4;
+    jacobian[i_cell][4][0] = 78.5;
+    jacobian[i_cell][4][2] = 53.6;
+    jacobian[i_cell][4][4] = 1.0;
+  }
+  
+  solver.AlphaMinusJacobian(jacobian, 42.042);
+  for (std::size_t i_cell = 0; i_cell < number_of_grid_cells; ++i_cell)
+  {
+    EXPECT_NEAR(jacobian[i_cell][0][0], 42.042 - 12.2, 1.0e-5);
+    EXPECT_NEAR(jacobian[i_cell][0][1], -24.3 * (i_cell + 2), 1.0e-5);
+    EXPECT_NEAR(jacobian[i_cell][0][2], -42.3, 1.0e-5);
+    EXPECT_NEAR(jacobian[i_cell][1][0], -0.43, 1.0e-5);
+    EXPECT_NEAR(jacobian[i_cell][1][1], 42.042 - 23.4, 1.0e-5);
+    EXPECT_NEAR(jacobian[i_cell][1][2], -83.4 / (i_cell + 3), 1.0e-5);
+    EXPECT_NEAR(jacobian[i_cell][2][0], -4.74, 1.0e-5);
+    EXPECT_NEAR(jacobian[i_cell][2][2], 42.042 - 6.91, 1.0e-5);
+    EXPECT_NEAR(jacobian[i_cell][3][1], -59.1, 1.0e-5);
+    EXPECT_NEAR(jacobian[i_cell][3][3], 42.042 - 83.4, 1.0e-5);
+    EXPECT_NEAR(jacobian[i_cell][4][0], -78.5, 1.0e-5);
+    EXPECT_NEAR(jacobian[i_cell][4][2], -53.6, 1.0e-5);
+    EXPECT_NEAR(jacobian[i_cell][4][4], 42.042 - 1.0, 1.0e-5);
```

Comment:
Why we can not use `EXPECT_EQ` here?

Reply:
These are the testings that C++ code used to check jacobian values after the function call. I am not sure if there's a loss after calculation in the algorithm. Maybe @mattldawson can explain this better. On the last part of the testing, I made sure the jacobian values from CUDA is equal to the jacobian values from C++ side.

Reply:
Actually I bet we could make these equal

Reply:
tests passed with EXPECT_EQ. Changes made.

Reply:
Thanks @qinatan .